// vue-use-fixed-header/dist/index.js
var K = Object.defineProperty, Q = Object.defineProperties;
var W = Object.getOwnPropertyDescriptors;
var M = Object.getOwnPropertySymbols;
var X = Object.prototype.hasOwnProperty, Z = Object.prototype.propertyIsEnumerable;
var F = (n, t, r) => t in n ? K(n, t, { enumerable: !0, configurable: !0, writable: !0, value: r }) : n[t] = r, a = (n, t) => {
  for (var r in t || (t = {}))
    X.call(t, r) && F(n, r, t[r]);
  if (M)
    for (var r of M(t))
      Z.call(t, r) && F(n, r, t[r]);
  return n;
}, S = (n, t) => Q(n, W(t));
import { ref as C, onMounted as q, onBeforeUnmount as ee, shallowRef as ne, watch as te, unref as o, readonly as ie, computed as P } from "vue";
const L = typeof window != "undefined";
function re() {
  const n = C(!1);
  if (!L)
    return n;
  const t = window.matchMedia("(prefers-reduced-motion: reduce)"), r = () => n.value = t.matches;
  return q(() => {
    var l;
    r(), (l = t.addEventListener) == null || l.call(t, "change", r);
  }), ee(() => {
    var l;
    (l = t.removeEventListener) == null || l.call(t, "change", r);
  }), n;
}
const B = "cubic-bezier(0.16, 1, 0.3, 1)", oe = {
  enterStyles: {
    transition: `all 0.35s ${B} 0s`,
    transform: "translateY(0px)"
  },
  leaveStyles: {
    transition: `all 0.5s ${B} 0s`,
    transform: "translateY(-101%)"
  }
};
function ce(n, t = {}) {
  const { enterStyles: r, leaveStyles: l } = oe, Y = re(), s = {
    resizeObserver: void 0,
    initResizeObserver: !1,
    isListeningScroll: !1,
    isHovering: !1
  }, v = ne({}), u = C(
    0
    /* READY */
  ), m = (e) => v.value = e, b = () => v.value = {}, y = (e) => u.value = e;
  function d() {
    const e = o(t.root);
    return e != null ? e : document.documentElement;
  }
  const h = () => d().scrollTop;
  function g() {
    const e = d();
    return e === document.documentElement ? document : e;
  }
  function _() {
    const e = o(n);
    if (!e)
      return !1;
    const { position: i, display: c } = window.getComputedStyle(e);
    return (i === "fixed" || i === "sticky") && c !== "none";
  }
  function A() {
    const e = o(n);
    if (!e)
      return 0;
    let i = e.scrollHeight;
    const { marginTop: c, marginBottom: f } = window.getComputedStyle(e);
    return i += Number.parseFloat(c) + Number.parseFloat(f), i;
  }
  function U() {
    s.resizeObserver = new ResizeObserver(() => {
      if (!s.initResizeObserver)
        return s.initResizeObserver = !0;
      H();
    }), s.resizeObserver.observe(d());
  }
  function w() {
    u.value !== 1 && (R(), m(S(a(a({}, r), o(t.transitionOpacity) ? { opacity: 1 } : {}), {
      visibility: ""
    })), y(
      1
      /* ENTER */
    ));
  }
  function V() {
    u.value !== 2 && (m(a(a({}, l), o(t.transitionOpacity) ? { opacity: 0 } : {})), y(
      2
      /* LEAVE */
    ), $());
  }
  function E(e) {
    if (R(), !o(n) || e.target !== o(n) || e.propertyName !== "transform")
      return;
    const { transform: i } = window.getComputedStyle(o(n));
    i !== "matrix(1, 0, 0, 1, 0, 0)" && m(S(a({}, l), {
      visibility: "hidden"
    }));
  }
  function $() {
    const e = o(n);
    e && e.addEventListener("transitionend", E);
  }
  function R() {
    const e = o(n);
    e && e.removeEventListener("transitionend", E);
  }
  function k() {
    let e = L ? h() : 0;
    return () => {
      const i = h(), c = i <= A(), f = i < e, p = i > e, N = Math.abs(i - e);
      if (c)
        return w();
      N < 10 || (s.isHovering || (f ? w() : p && V()), e = i);
    };
  }
  const O = k();
  function D() {
    g().addEventListener("scroll", O, { passive: !0 }), s.isListeningScroll = !0;
  }
  function G() {
    g().removeEventListener("scroll", O), s.isListeningScroll = !1;
  }
  const T = () => s.isHovering = !0, z = () => s.isHovering = !1;
  function j() {
    var e, i;
    (e = o(n)) == null || e.addEventListener("pointerenter", T), (i = o(n)) == null || i.addEventListener("pointerleave", z);
  }
  function J() {
    var e, i;
    (e = o(n)) == null || e.removeEventListener("pointerenter", T), (i = o(n)) == null || i.removeEventListener("pointerleave", z);
  }
  function H() {
    const e = _();
    s.isListeningScroll ? e || (x(), b()) : e && (D(), j());
  }
  function x() {
    G(), J();
  }
  return L && te(
    () => [o(n), d(), Y.value, o(t.watch)],
    ([e, i, c], f, p) => {
      !c && e && (i || i === null) && (U(), H()), p(() => {
        var I;
        x(), b(), (I = s.resizeObserver) == null || I.disconnect(), s.initResizeObserver = !1;
      });
    },
    { immediate: !0, flush: "post" }
  ), {
    styles: ie(v),
    isLeave: P(
      () => u.value === 2
      /* LEAVE */
    ),
    isEnter: P(
      () => u.value === 1
      /* ENTER */
    )
  };
}
export {
  ce as useFixedHeader
};
